qui/updater/progress_page.py

```diff
@@ -0,0 +1,438 @@
+        self.progressbar_renderer: Gtk.CellRendererProgress = \
+            self.builder.get_object("progressbar_renderer")
+        self.progressbar_renderer.set_fixed_size(-1, 26)
+
+        self.progress_list: Gtk.TreeView = self.builder.get_object(
+            "progress_list")
+        self.selection: Gtk.TreeSelection = self.progress_list.get_selection()
+        self.progress_list.connect("row-activated", self.row_selected)
+        progress_column: Gtk.TreeViewColumn = self.builder.get_object(
+            "progress_column")
+        renderer = CellRendererProgressWithResult()
+        renderer.props.ypad = 10
+        progress_column.pack_start(renderer, True)
+        progress_column.add_attribute(renderer, "pulse", 7)
+        progress_column.add_attribute(renderer, "value", 7)
+        progress_column.add_attribute(renderer, "status", 8)
+
+    @property
+    def is_visible(self):
+        """Returns True if page is shown by stack."""
+        return self.stack.get_visible_child() == self.page
+
+    def init_update(self, vms_to_update, settings):
+        """Starts `perform_update` in new thread."""
+        self.vms_to_update = vms_to_update
+        self.progress_list.set_model(vms_to_update.list_store_raw)
+        self.next_button.set_sensitive(False)
+        self.cancel_button.set_sensitive(True)
+        self.cancel_button.set_label(l("_Cancel updates"))
+        self.cancel_button.show()
+
+        self.header_label.set_text(l("Update in progress..."))
+        self.header_label.set_halign(Gtk.Align.CENTER)
+
+        self.update_thread = threading.Thread(
+            target=self.perform_update,
+            args=(settings,)
+        )
+        self.update_thread.start()
+
+    def perform_update(self, settings):
+        """Updates dom0 and then other vms."""
+        admins = [row for row in self.vms_to_update
+                  if row.vm.klass == 'AdminVM']
+        templs = [row for row in self.vms_to_update
+                  if row.vm.klass != 'AdminVM']
+
+        if admins:
+            self.update_admin_vm(admins)
+
+        if templs:
+            self.update_templates(templs, settings)
+
+        GObject.idle_add(self.next_button.set_sensitive, True)
+        GObject.idle_add(self.header_label.set_text, l("Update finished"))
+        GObject.idle_add(self.cancel_button.set_visible, False)
+
+    def update_admin_vm(self, admins):
+        """Runs command to update dom0."""
+        admin = admins[0]
+        if self.exit_triggered:
+            GObject.idle_add(admin.set_status, UpdateStatus.Cancelled)
+            GObject.idle_add(
+                admin.append_text_view,
+                l("Canceled update for {}\n").format(admin.vm.name))
+
+        GObject.idle_add(
+            admin.append_text_view,
+            l("Updating {}\n").format(admin.name))
+        GObject.idle_add(admin.set_status, UpdateStatus.ProgressUnknown)
+
+        self.update_details.update_buffer()
```

Comment:
I've got the following in details of updating a fedora-37: `b'Generating grub configuration file...` (and so on) - looks like some binary text doesn't decode.